library/oid.c

```diff
@@ -834,21 +834,55 @@ int mbedtls_oid_get_numeric_string(char *buf, size_t size,
     p = buf;
     n = size;
 
-    /* First byte contains first two dots */
-    if (oid->len > 0) {
-        ret = mbedtls_snprintf(p, n, "%d.%d", oid->p[0] / 40, oid->p[0] % 40);
-        OID_SAFE_SNPRINTF;
+    /* First subidentifier contains first two OID components */
+    i = 0;
+    value = 0;
+    if ((oid->p[0]) == 0x80) {
+        /* Overlong encoding is not allowed */
+        return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
```

Comment:
I agree that `BUF_TOO_SMALL` is not the right error here. `BUF_TOO_SMALL` means “this would work with a larger buffer” (unless some other error also applies). The only other OID error code is `OID_NOT_FOUND` which is distinctive and wouldn't fit here. We need a `BAD_INPUT_DATA` error for OID.

Regarding error code assignments: see [`include/mbedtls/error.h`](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/error.h#L30). Beware that the list is often out-of-date. To find free low-level error codes:
```
scripts/config.py full
make
for i in 0 1; do for j in {1..63}; do programs/util/strerror $((i+2*j)); done; done | grep 'UNKNOWN ERROR'
```

OID is a low-level module so its error codes must be in the range 0x02–0x7e. We try to give each module one range of either odd or even codes (I have no idea why they're separated), but the relatively small size of the space has caused a lot of irregularities. 0x0030 is free and next to 0x002E which is allocated to OID, so 0x0030 would be a good value.

Alternatively, we could create a generic `BAD_INPUT_DATA` error code, like we've done for `GENERIC_ERROR` and `CORRUPTION_DETECTED`. When we worked on Mbed TLS 3.0, we had a consensus that it was desirable to unify same-meaning error codes across modules, but we had limited time and couldn't do implement it. I hope we'll do that in 4.0 (building on `psa_status_t`). But I think it makes the API more cumbersome to do this only for new modules. As a precedent, we recently added error codes for LMS.

Reply:
From `error.h`:

```
 * Module   Nr  Codes assigned
 * OID       1  0x002E-0x002E   0x000B-0x000B
 * PADLOCK   1  0x0030-0x0030
```

and

```
library/padlock.h:#define MBEDTLS_ERR_PADLOCK_DATA_MISALIGNED               -0x0030  /**< Input data should be aligned. */
```

Since we haven't got rid of the VIA yet, I don't think we can use `-0x0030`.

> We try to give each module one range of either odd or even codes (I have no idea why they're separated)

While `-0x002D` would break this rule, if we don't know why we used to do it, and since OID already uses `-0x000B`, why not break (the rule) again for `MBEDTLS_ERR_OID_INVALID` of `-0x002D`?

Alternatively I suggest `-0x0028 - UNKNOWN ERROR CODE (0028)` for `MBEDTLS_ERR_INVALID_INPUT_DATA` - thoughts?

Reply:
Well, that's a problem: error codes need to be defined in the public API. Otherwise, users can't check for the error! (Except by breaking the abstraction boundary.)

Except that I'm not sure that `MBEDTLS_ERR_PADLOCK_DATA_MISALIGNED` can actually be returned to applications. The padlock functions aren't part of the API, and I suspect that the code in `aes.c` only passes correctly aligned data, and so the error is never raised.

[To be checked](https://github.com/Mbed-TLS/mbedtls/issues/7118).

If we can't use 0x0030, we have to pick a third (single-element) range. Messy, but won't be the first time.

Reply:
I still don't understand why we can't use `-0x002D`? To be sure, it's not an even number, but OID already uses `-0x000B` (ironically it's `MBEDTLS_ERR_OID_BUF_TOO_SMALL`, the one that is being painfully overloaded here)

Reply:
Sure, we can use 0x002D. That's a distinct range from 0x002E-0x002E and 0x000B-0x000B since odd values and even values are considerd independently.

Reply:
Ah, there is also `MBEDTLS_ERR_ASN1_INVALID_DATA`, which might be better

Reply:
Replaced with more precise errors borrowed from the ASN.1 module.